### ISSUE TITLE
Do you multiply by the masks in the forward pass?

### ISSUE THREAD

**Issue by rolson24** [Poirazi-Lab/Chavlis_Poirazi_2024#2]
In this part of the paper you say that the weights in the dANN models are masked off:
```
We have developed a range of traditional ANNs consisting of two hidden layers and an output layer that matches the 
number of classes. To create the dANN model, we first create two boolean masks that determine the synaptic weights
between the input and dendritic layers, as well as the cable weights between dendritic and somatic layers. Once we
have initialized the model, we apply these masks to achieve a sparse network with structured synaptic and cable 
weights (Eq. [1](https://www.nature.com/articles/s41467-025-56297-9#Equ1)).

where W_k denotes the weights, M_k the boolean mask associated k-th layer, and (*) is the Hadamard (elementwise) product.
```

However in your code it seems like you build a typical dense neural network and just do the forward pass on the dense network:
```
    @tf.function
    def train_step(x, y):
        with tf.GradientTape() as tape:
            # Run the forward pass of the layer.
            # Logits for this minibatch
            logits = model(x, training=True)
            # Compute the loss value for this minibatch.
            loss_value = loss_fn(y, logits)
        # Use the gradient tape to automatically retrieve
        # the gradients of the trainable variables with respect to the loss.
        grads = tape.gradient(loss_value, model.trainable_weights)
        # Apply the masks to zero out gradients of non existing connections.
        grads_masked = [tf.math.multiply(g, m) for g, m in zip(grads, Masks)]
        # Check that updated gradients shape is the same as gradients.
        if len(grads) != len(grads_masked):
            raise ValueError("Gradients are unequal in size after masking.")
        # Run one step of gradient descent by updating the value of the
        # variables to minimize the loss.
        optimizer.apply_gradients(zip(grads_masked, model.trainable_weights))
        # Update the training metrics.
        train_acc_metric.update_state(y, logits)
        train_loss_metric.update_state(y, logits)
        return loss_value
```

I can see that you mask out the gradients in the backward pass which should mean the masked weights won't get trained, however they are still in the model during the forward pass so they are still impacting the model because they allow a layer to still "see" all of the neurons in the previous layer which allows the model to still use the information from these connections even if they cannot modify the weights.

I am trying to reproduce the results using sparse matrices for the forward pass, however from some initial tests this difference is causing the model to not perform as well as your model. This suggests that the un-trainable params are still providing a significant advantage to the model.

**Comment by edmondja** [comment]
Yes for me as well it is not so clear. In the paper it is said that weights are zeroed-out after each training step, but then what is the point of also zeroing-out the gradients for those weights ?

**Comment by spirosChv** [comment]
> I can see that you mask out the gradients in the backward pass which should mean the masked weights won't get trained, however they are still in the model during the forward pass so they are still impacting the model because they allow a layer to still "see" all of the neurons in the previous layer which allows the model to still use the information from these connections even if they cannot modify the weights.


Thank you for your comment. After the model initialization, the weights are multiplied by the Masks (as detailed in the code [here](https://github.com/Poirazi-Lab/dendritic_anns/blob/c0d6617ecace0d075ccf5d4914bfc2cf410eff68/codes_/main.py#L176C1-L181C29)).

During the first forward pass, the model utilizes only the sparse weights. Following this, gradients are computed, and we zero out all gradients that do not correspond to the connections defined by the initial network. This technique ensures that the initial weights remain unchanged throughout the training process. The gradient masking happens after each backdrop step.

I trust this clarifies your inquiry. If you have further questions or would like to discuss this in detail, please feel free to reach out via email.


> 
> I am trying to reproduce the results using sparse matrices for the forward pass, however from some initial tests this difference is causing the model to not perform as well as your model. This suggests that the un-trainable params are still providing a significant advantage to the model.

I do not understand how the zero weights impact the training if they remain zero in every forward pass.

**Comment by spirosChv** [comment]
> Yes for me as well it is not so clear. In the paper it is said that weights are zeroed-out after each training step, but then what is the point of also zeroing-out the gradients for those weights ?

Starting from a sparse network, after the backpropagation step, we indeed obtain gradients (Δw) in positions where connections do not exist. Given that the weight update follows the rule
```python
w-=ηΔw
```
this can lead to the creation of weights in previously unconnected positions, which is inconsistent with biological principles.

To address this, our approach involves zeroing out gradients for those connections that were not part of the initial sparse network. This ensures that only the weights corresponding to existing connections are updated during training, thereby aligning our model more closely with biological reality.

If you have any further questions or need additional clarification, please don't hesitate to reach out.
